components/ws-manager/pkg/manager/create_test.go

```diff
@@ -177,3 +177,124 @@ func TestCreateDefiniteWorkspacePod(t *testing.T) {
+		PVCConfig        *config.PVCConfiguration      `json:"pvcConfig,omitempty"`
+		ResourceRequests *config.ResourceConfiguration `json:"resourceRequests,omitempty"`
+		ResourceLimits   *config.ResourceConfiguration `json:"resourceLimits,omitempty"`
+	}
+	type fixture struct {
+		WorkspaceClass
+
+		Spec         *json.RawMessage          `json:"spec,omitempty"`    // *api.StartWorkspaceSpec
+		Request      *json.RawMessage          `json:"request,omitempty"` // *api.StartWorkspaceRequest
+		Context      *startWorkspaceContext    `json:"context,omitempty"`
+		CACertSecret string                    `json:"caCertSecret,omitempty"`
+		Classes      map[string]WorkspaceClass `json:"classes,omitempty"`
+
+		EnforceAffinity bool `json:"enforceAffinity,omitempty"`
+	}
+	type gold struct {
+		PVC   corev1.PersistentVolumeClaim `json:"reason,omitempty"`
+		Error string                       `json:"error,omitempty"`
+	}
+
+	test := ctesting.FixtureTest{
+		T:    t,
+		Path: "testdata/cpwp_*.json",
+		Test: func(t *testing.T, input interface{}) interface{} {
+			fixture := input.(*fixture)
+
+			mgmtCfg := forTestingOnlyManagerConfig()
+			mgmtCfg.WorkspaceCACertSecret = fixture.CACertSecret
+
+			if fixture.Classes == nil {
+				fixture.Classes = make(map[string]WorkspaceClass)
+			}
+
+			if _, exists := fixture.Classes[config.DefaultWorkspaceClass]; !exists {
+				if fixture.WorkspaceClass.ResourceLimits != nil || fixture.WorkspaceClass.ResourceRequests != nil {
+					// there's no default class in the fixture. If there are limits configured, use those
+					fixture.Classes[config.DefaultWorkspaceClass] = fixture.WorkspaceClass
+				}
+			}
+
+			for n, cls := range fixture.Classes {
+				var cfgCls config.WorkspaceClass
+				cfgCls.Container.Requests = cls.ResourceRequests
+				cfgCls.Container.Limits = cls.ResourceLimits
+				if cls.PVCConfig != nil {
+					cfgCls.PVC = *cls.PVCConfig
+				}
+
+				mgmtCfg.WorkspaceClasses[n] = &cfgCls
+			}
+
+			manager := &Manager{Config: mgmtCfg}
+
+			if fixture.Context == nil {
+				var req api.StartWorkspaceRequest
+				if fixture.Request == nil {
+					if fixture.Spec == nil {
+						t.Errorf("fixture has neither context, nor request, nor spec")
+						return nil
+					}
+
+					var spec api.StartWorkspaceSpec
+					err := protojson.Unmarshal([]byte(*fixture.Spec), &spec)
+					if err != nil {
+						t.Errorf("cannot unmarshal StartWorkspaceSpec: %v", err)
+						return nil
+					}
+
+					req = api.StartWorkspaceRequest{
+						Type: api.WorkspaceType_REGULAR,
+						Id:   "test",
+						Metadata: &api.WorkspaceMetadata{
+							Owner:  "tester",
+							MetaId: "foobar",
+						},
+						ServicePrefix: "foobarservice",
+						Spec:          &spec,
+					}
+				} else {
+					err := protojson.Unmarshal([]byte(*fixture.Request), &req)
+					if err != nil {
+						t.Errorf("cannot unmarshal StartWorkspaceReq: %v", err)
+						return nil
+					}
+				}
+
+				if req.Spec.Class == "" {
+					fmt.Println()
+				}
```

Comment:
Any reason to print this empty line?

Reply:
copy paste error from the function above. :(